### PR TITLE
Consider confidence values in 'ensure_consistent_bilou_tagging'

### PR DESCRIPTION
**Pull request by tabergma**
**Proposed changes**:
The method `ensure_consistent_bilou_tagging` now also considers the confidence values of the predicted tags
when updating the BIlOU tags.

closes https://github.com/RasaHQ/rasa/issues/6039

**Status (please check what you already did)**:
- [x] added some tests for the functionality
- [ ] updated the documentation
- [x] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/master/changelog) for instructions)
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

File: rasa/nlu/utils/bilou_utils.py [hunk 1]
```diff
@@ -271,6 +277,32 @@ def ensure_consistent_bilou_tagging(predicted_tags: List[Text]) -> List[Text]:
...
+            # if not all tags are the same, for example, B-person I-person L-location
+            # we need to check what tag we should use depending on the confidence
+            # values and update the tags and confidences accordingly
+            if not all(relevant_tags[0] == tag for tag in relevant_tags):
+                avg_confidence_per_tag = _avg_confidence_per_tag(
+                    relevant_tags, relevant_confidences
+                )
+
+                # the entity should have the tag with the highest average confidence
+                tag = max(avg_confidence_per_tag, key=avg_confidence_per_tag.get)
```
Comment (tabergma) [inline rasa/nlu/utils/bilou_utils.py:294]: @Ghostvv do you think this is enough or should we also take into account how many tags within the entity are predicted as e.g. person or location?

Comment (Ghostvv) [reply rasa/nlu/utils/bilou_utils.py:294]: I think how many tags is more robust measure. E.g. person 2 times with 0.89 and location 1 time with 0.95. Maybe it is possible to come up with some kind of F1 score between these values

Comment (Ghostvv) [reply rasa/nlu/utils/bilou_utils.py:294]: also, it'd be great to log this mistakes somehow, so that users can be pointed out where more data is required

Comment (tabergma) [reply rasa/nlu/utils/bilou_utils.py:294]: 👍

Comment (tabergma) [reply rasa/nlu/utils/bilou_utils.py]: updated the metric

File: rasa/nlu/utils/bilou_utils.py [hunk 2]
```diff
@@ -283,10 +311,113 @@ def ensure_consistent_bilou_tagging(predicted_tags: List[Text]) -> List[Text]:
...
+            2
+            * (avg_confidence * token_percentage)
+            / (avg_confidence + token_percentage)
+        )
+
+    # Take the tag with the highest score as the tag for the entity
+    tag = max(score_per_tag, key=score_per_tag.get)
+    avg_confidence = avg_confidence_per_tag[tag]
+
+    return tag, avg_confidence
```
Comment (Ghostvv) [inline rasa/nlu/utils/bilou_utils.py:357]: should the score be used as "confidence" then?

**Review by Ghostvv — COMMENTED** [review]
not sure how troublesome to implement it, but is it possible to output original tags during `rasa test` as well?

**Comment by tabergma** [comment]
You mean the tags before they got updated? mmhh... not that easy, as the tags are converted into proper entities and I am not sure how you would add the "old" tags in there.

File: rasa/nlu/utils/bilou_utils.py (hunk 2 above)
Comment (tabergma) [reply rasa/nlu/utils/bilou_utils.py:357]: might make sense, the score would be for sure lower compared to the avg confidence score

Comment (Ghostvv) [reply rasa/nlu/utils/bilou_utils.py]: but that's good I guess, since it is kind of artificial substitution

**Comment by Ghostvv** [comment]
Let's create an issue for it to discuss it. So that this PR is not blocked

**Comment by tabergma** [comment]
https://github.com/RasaHQ/rasa/issues/6647

**Review by Ghostvv — APPROVED** [review]

**Comment by Ghostvv** [comment]
Could you please add the description of this substitution process to docs, please?

**Comment by tabergma** [comment]
Where should it go? We do not have a section for it as far as i know.

**Comment by Ghostvv** [comment]
hm, it seems that we don't have a place where we describe bilou tagging scheme. I think we should add it some where

**Comment by tabergma** [comment]
@erohmensing tagging you as a review as I added a section about BILOU tagging to the docs. As the information about `NLU Training Data` was already included in `Training Data Format` I removed it and kept just the entity related stuff. Let me know if that does not work and the section should be moved somewhere else.

**Review by Ghostvv — COMMENTED** [review]
the part about BILOU tags looks good.
I wouldn't remove `nlu-training-data.mdx` as part of this PR. To avoid confusion, let's do it in a separate PR

**Comment by tabergma** [comment]
followed the suggestion of @Ghostvv and added the page about nlu training data again and just added the bilou section to that page

File: changelog/6039.improvment.md [hunk 3]
```diff
@@ -0,0 +1,2 @@
+The method `ensure_consistent_bilou_tagging` now also considers the confidence values of the predicted tags
+when updating the BIlOU tags.
```
Comment (erohmensing) [inline changelog/6039.improvment.md:2]:
```suggestion
when updating the BILOU tags.
```

File: docs/docs/nlu-training-data.mdx [hunk 4]
```diff
@@ -168,3 +168,41 @@ For example, you should not only have example like `fly FROM x TO y`, but also i
 
 To fill slots from entities with a specific role/group, you need to either define a custom slot mappings using
 [Forms](forms.mdx) or use [Custom Actions](custom-actions.mdx) to extract the corresponding entity directly from the tracker.
+
+
+### Entity with BILOU tagging
```
Comment (erohmensing) [inline docs/docs/nlu-training-data.mdx:173]: Placement is fine for now, i'll find a spot with it when reviewing your updates to this page :) 
```suggestion
### BILOU Entity Tagging
```

File: docs/docs/nlu-training-data.mdx [hunk 5]
```diff
@@ -168,3 +168,41 @@ For example, you should not only have example like `fly FROM x TO y`, but also i
...
+
+[DIETClassifier](./components.mdx#dietclassifier) and [CRFEntityExtractor](./components.mdx#crfentityextractor)
+have the option `BILOU_flag`.
+`BILOU` is short for `Beginning, Inside, Last, Outside, and Unit-length` and refers to a tagging schema that can be
+used by the machine learning model.
```
Comment (erohmensing) [inline docs/docs/nlu-training-data.mdx:178]:
```suggestion
The [DIETClassifier](./components.mdx#dietclassifier) and [CRFEntityExtractor](./components.mdx#crfentityextractor)
have the option `BILOU_flag`, which refers to a tagging schema that can be
used by the machine learning model when processing entities.
`BILOU` is short for Beginning, Inside, Last, Outside, and Unit-length.

```

File: docs/docs/nlu-training-data.mdx [hunk 6]
```diff
@@ -168,3 +168,41 @@ For example, you should not only have example like `fly FROM x TO y`, but also i
...
+For example, the training example
+
+```
+[Alex]{"entity": "person"} is going with [Marty A. Rick]{"entity": "person"} to [Los Angeles]{"entity": "location"}.
+```
+
+is first split into a list of tokens and then the machine learning model applies the tagging schema
+depending on the option `BILOU_flag` during training.
```
Comment (erohmensing) [inline docs/docs/nlu-training-data.mdx:186]:
```suggestion
is first split into a list of tokens. Then the machine learning model applies the tagging schema
as shown below, depending on the value the `BILOU_flag`:
```

File: docs/docs/nlu-training-data.mdx [hunk 7]
```diff
@@ -168,3 +168,41 @@ For example, you should not only have example like `fly FROM x TO y`, but also i
...
+
+| token   | BILOU tagging schema | normal tagging schema |
```
Comment (erohmensing) [inline docs/docs/nlu-training-data.mdx:188]:
```suggestion
| token   | `BILOU_flag = true`  | `BILOU_flag = false`  |
```

File: docs/docs/nlu-training-data.mdx [hunk 8]
```diff
@@ -168,3 +168,41 @@ For example, you should not only have example like `fly FROM x TO y`, but also i
...
+| rick    | L-person             | person                |
+| to      | O                    | O                     |
+| los     | B-location           | location              |
+| angeles | L-location           | location              |
+
+The BILOU tagging schema is richer compared to the normal tagging schema. It may help to improve the
+performance of the machine learning model when predicting entities.
+
+:::note
+When the option `BILOU_flag` is set to `True`, it might happen that the model is predicting inconsistent BILOU tags, e.g.
```
Comment (erohmensing) [inline docs/docs/nlu-training-data.mdx:205]:
```suggestion
When the option `BILOU_flag` is set to `True`, the model may predict inconsistent BILOU tags, e.g.
```

File: docs/docs/nlu-training-data.mdx [hunk 9]
```diff
@@ -168,3 +168,41 @@ For example, you should not only have example like `fly FROM x TO y`, but also i
...
+| a       | I-person             | person                |
+| rick    | L-person             | person                |
+| to      | O                    | O                     |
+| los     | B-location           | location              |
+| angeles | L-location           | location              |
+
+The BILOU tagging schema is richer compared to the normal tagging schema. It may help to improve the
+performance of the machine learning model when predicting entities.
+
+:::note
```
Comment (erohmensing) [inline docs/docs/nlu-training-data.mdx:204]:
```suggestion
:::note inconsistent bilou tags
```

**Review by erohmensing — COMMENTED** [review]
👍

**Comment by tabergma** [comment]
Applied all review comments, thanks for the comments!

@erohmensing or @Ghostvv can someone please approve so that we can merge this PR? Thanks.

**Review by erohmensing — APPROVED** [review]

**Comment by tabergma** [comment]
oh no, just realised that we will not be able to merge as a lot of tests will all fail due to the convert issue

**Pull request merged**